### ISSUE TITLE
Concerning the error reported after training for 98 epochs, it indicates that there is not enough GPU memory.

### ISSUE THREAD

**Issue by Fackyhub** [ultralytics/ultralytics#12679]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I use Python for training, the following are the training parameters：rect=True，device=[0,1]，workers=0，batch=4。My images size H(3648) and W(5472), but each one is the same size.error reported after training for 98 epochs, it indicates that there is **cuda of memory**.I don't understand why if the memory is not enough, it doesn't give an error in the initial few epochs, but this issue appears after training 98 epochs.   My GPU is rtx4090（24G） * 2.

### Additional

 I have also noticed the previous question, rectangular training is currently incompatible with multi-GPU，but it still work，This warning doesn't seem to affect the training.

**Comment by glenn-jocher** [comment]
@Fackyhub hello! It seems like you're encountering a GPU memory issue deep into your training process. This can sometimes happen due to accumulating gradients or other subtleties in memory management that don't manifest until later epochs.

Here are a couple of suggestions to mitigate this issue:

1. **Reduce Batch Size**: Try lowering the `batch` size to reduce memory consumption per step.
2. **Use Gradient Accumulation**: If reducing the batch size impacts model performance, consider implementing gradient accumulation to effectively increase the batch size without increasing memory usage.

Here's a quick example of how you might implement gradient accumulation:

```python
accumulation_steps = 4  # Number of steps to accumulate gradients over
for i, (inputs, labels) in enumerate(data_loader):
    predictions = model(inputs)
    loss = criterion(predictions, labels)
    loss = loss / accumulation_steps  # Normalize the loss
    loss.backward()  # Accumulate gradients
    if (i + 1) % accumulation_steps == 0:
        optimizer.step()  # Perform a real update
        optimizer.zero_grad()  # Reset gradients after update
```

This approach allows you to effectively train with a larger batch size by accumulating gradients over several iterations, thus reducing the GPU memory required per iteration.

Also, keep an eye on any other processes that might be consuming GPU memory. Sometimes, freeing up resources or restarting the system might help clear any lingering memory usage.

Hope this helps! Let me know if you have any more questions. 😊

**Comment by Fackyhub** [comment]
@glenn-jocher Thanks，The batch-size is already very small，So I tried to Use Gradient Accumulation.Then I saw this line of code in the program---self.accumulate = max(round(self.args.nbs / self.batch_size), 1)，it'work.That indicates that the original code will perform gradient accumulation。By the way,I found that it doesn't work when rect=true during training，Because of this line of code---return build_yolo_dataset(self.args, img_path, batch, self.data, mode=mode, rect=mode == "val", stride=gs)，rect=mode == "val" is default.so training it doesn't word.I have modified it.

**Comment by glenn-jocher** [comment]
Hi @Fackyhub, thanks for the update! It looks like you've made some insightful observations about the gradient accumulation and the `rect` parameter behavior during training. Good catch on the `rect=mode == "val"` condition—it's indeed set to work only during validation by default. Modifying it for training as you did can help utilize rectangular training if that's what your dataset requires. If you encounter any more quirks or need further assistance tweaking the settings, feel free to reach out. Happy training! 😊
